packages/migrate/src/transforms/2345-fix-transform.ts

```diff
@@ -0,0 +1,28 @@
+import ts from 'typescript';
+
+import FixTransform, { FixedFile } from '../interfaces/fix-transform';
+import { findNodeAtPosition } from '../helpers/typescript-ast';
+
+export default class FixTransform2345 extends FixTransform {
+  hint = `Argument of type '{0}' is not assignable to parameter of type '{1}'.`;
+
+  fix = (diagnostic: ts.DiagnosticWithLocation): FixedFile[] => {
+    const errorNode = findNodeAtPosition(diagnostic.file, diagnostic.start, diagnostic.length);
+    const variableName = errorNode?.getText().trim();
+    const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, '. ');
+    const types = message.match(/'([^']+)'/gm)?.map((matched) => matched.replace(/^'|'$/g, ''));
+    if (Array.isArray(types) && variableName) {
+      const [actualType, expectedType] = types;
+      if (actualType === 'unknown' || actualType === 'any') {
```

Comment:
If this use case goes beyond the current implementation of [prepareHint](https://github.com/rehearsal-js/rehearsal-js/blob/948eb8a8dea1aae76f26d82e4cf39f919ffd68c4/packages/migrate/src/plugins/diagnostics-autofix.plugin.ts#L68) we should consider expanding the method to handle this use case. 

Or just do a positional replace on the delta of the help message. Looks like the delta is only `specifying type of argument to be ${expectedType}`. eg

hint = `Argument of type '{0}' is not assignable to parameter of type '{1}'. '{2}', using type assertion: '(${variableName} as ${expectedType})', or using type guard: 'if (${variableName} instanceof ${expectedType}) { ... }'.`

Reply:
Hint intentionally left simple to keep all conditions in fix function. For some condition fixes can be applied in a first place and for other - only hints. Later hints can be replaced with auto fixes, but no conditions will be changed

Reply:
@lynchbomb @lynchbomb My first version of TS2345 is to create the hint in the `fix()` method. I can see the merits of both approaches: merging the hint and the autofix in one `fix()` function or separate functions for each. 

The first approach may be for performance in certain cases when both autofix and hints share access to the diagnostic node. 

The second approach may be better for readability and maintenance. In addition, the fix() method returns the files that have been changed and the plugin file uses that value to report whether any autofix has been done. If the fix() is also responsible for producing hints, then it's a function that does two things at one time. And the need for shared access to the diagnostic node doesn't happen each time when creating hints and modifying AST. 

Each has tradeoffs. I am OK with either way.

Reply:
> The first approach may be for performance in certain cases when both autofix and hints share access to the diagnostic node.

There is no difference in performance in case you don't need a node for message (you don't required to traverse), but there is, actually, decrease performance [in case](https://github.com/rehearsal-js/rehearsal-js/pull/205/files#diff-224757bec7f6966cf22692d794cec47df113b520f11fc1b6dddb9df97712ecc6R8-R35) of having both fix (for some cases) and hint (for other) because you need to traverse tree twice - first in fix(), second in getHint().

Since, the target is to provide better hints (even with some code snippets), the node [is usually needed](https://github.com/rehearsal-js/rehearsal-js/pull/205/files#diff-c0bb29af946a24a72228039bfe46599f98193e72d82154391159b2997b10426aL18). So, we need to traverse anyway. 

> The second approach may be better for readability ....

Yeah, this is known issue, but there is a plan to address it a different way.
Renaming function fix() to something else to highlight this is not just a fix, but to be an entry point entry point / controller for autofix class, which will be called fixOneCase(), fixOtherCase(), ... or provideHint() depends on conditions.

> and maintenance.

For this purpose fix should only have condition and either call appropriate fix function or change hint (or, probably,  even both in future.

Like:
```
const text = transformDiagnosedNode(diagnostic, (node: ts.Node) => {
      if (ts.isImportDeclaration(node) || ts.isImportSpecifier(node)) {
        this.removeImportDeclaration()
      } else if (ts.isIdentifier(node) && ts.isFunctionDeclaration(node.parent)) {
        this.RepoveFunctionDeclaration()
      } else if (ts.isIdentifier(node) && ts.isParameter(node.parent)) {
        this.hint = `The function '{0}' is never called. Remove the function or use it.`;
        // Or even this.setHint(`The function '{0}' is never called. Remove the function or use it.`;)
      }

      return node;
    });
```

There is no benefits of using `findNodeAtPosition` over `transformDiagnosedNode` because both of them do the same, except transformDiagnosedNode fires callback while findNodeAtPosition you have to do perform additional checks for the node manually. 

> Each has tradeoffs. I am OK with either way.